hi-ml-cpath/src/health_cpath/preprocessing/loading.py

```diff
@@ -204,38 +212,127 @@ def __call__(self, data: Dict) -> Dict:
+    label from unique values in the mask if roi_label is None. Otherwise the sub ROI is sampled from the foreground of
+    the specified label.
+
+    Operates on dictionaries, replacing the file paths in `image_key` loaded array, in (C, H, W) format. Also adds the
+    following meta-data entries:
+    - `'location'` (tuple): top-right coordinates of the bounding box
+    - `'size'` (tuple): width and height of the bounding box
+    - `'level'` (int): chosen magnification level
+    - `'scale'` (float): corresponding scale, loaded from the file
+    """
+
+    def __init__(
+        self,
+        image_key: str = SlideKey.IMAGE,
+        mask_key: str = SlideKey.MASK,
+        wsi_mag_at_level0: float = 10.0,
+        wsi_mag_at_level: float = 10.0,
+        mask_mag: float = 1.25,
+        roi_label: Optional[Sequence[int]] = None,
+        **kwargs: Any,
+    ) -> None:
+        """
+        :param image_key: Image key in the input and output dictionaries. Default: 'image'.
+        :param mask_key: Mask key in the input and output dictionaries. Default: 'mask'.
+        :param wsi_mag_at_level0: WSI magnification at level 0. Default: 10x.
+        :param wsi_mag_at_level: WSI magnification at the chosen level. Default: 10x.
+        :param mask_mag: Mask magnification. Default: 1.25x.
+        :param roi_label: The label corresponding to the foreground tissue to sub sectionned. If None (default), a
+            random label is sampled from the foreground tissue uniformly.
+        :param kwargs: Additional arguments for `BaseLoadROId`.
+        """
+        MapTransform.__init__(self, [image_key, mask_key], allow_missing_keys=False)
+        BaseLoadROId.__init__(self, image_key=image_key, **kwargs)
+        self.mask_key = mask_key
+        self.scale_level0 = wsi_mag_at_level0 / mask_mag
+        self.scale_level = wsi_mag_at_level / (mask_mag * self.scale_level0)
+        self.roi_label = roi_label
+
+    def _get_sub_section_label(self, mask: np.ndarray) -> int:
+        """Return the label of the sub section to crop. If roi_label is None, a random label is sampled from the
+        foreground tissue uniformly. Otherwise the sub ROI is sampled from the foreground of the specified label."""
+        return self.roi_label or self.R.choice(np.unique(mask).tolist())
+
+    def _get_foreground_mask(self, mask_path: str, level: int = 0) -> np.ndarray:
+        """Return a binary foreground mask using the specified label or a random label if roi_label is None."""
+        mask = np.asarray(Image.open(mask_path))[..., 0]
+        section = self._get_sub_section_label(mask)
+        foreground_mask = mask == section
+        return foreground_mask
+
+    def _get_bounding_box(self, foreground_mask: np.ndarray, slide_id: Union[str, int] = "") -> box_utils.Box:
+        """Estimate bounding box at the lowest resolution (i.e. highest level) of the slide."""
```

Comment:
do you mean at level 0? Isn't this highest resolution?

Reply:
Oh sorry I didn't update the docs when I copied the function definition from above, will do!